qiime2/sdk/plugin_manager.py

```diff
@@ -142,34 +144,75 @@ def get_semantic_types(self):
 
         return types
 
+    class FormatFilters(enum.Flag):
+        EXPORTABLE = enum.auto()
+        IMPORTABLE = enum.auto()
+
     # TODO: Should plugin loading be transactional? i.e. if there's
     # something wrong, the entire plugin fails to load any piece, like a
     # databases rollback/commit
 
-    def get_formats(self, *, include_all=False, importable=False,
-                    exportable=False, canonical_format=False):
-        if include_all is True and canonical_format is True or include_all \
-                is True and importable is True or include_all is True and \
-                exportable is True:
-            raise ValueError("If all formats are requested, other formats "
-                             "cannot be included as a result.")
-
-        elif include_all is True:
-            return self.formats
+    def get_formats(self, *, filter=None, semantic_type=None):
 
         result_formats = set()
 
-        if importable is True:
-            result_formats = result_formats.union(self._importable)
+        if semantic_type is None:
```

Comment:
I think we should have a clearly documented "order of operations" for the filtering operations. My gut instinct is that the `semantic_type` filter should be applied _first_, which means that the logic could be cleaned up a bit. I think you can take a strategy of progressively filtering. Start with the semantic type param, and filter the list of formats as necessary. Then, taking that resulting (semantic-type-filtered) list of formats, apply any of the filter param filters, as appropriate. By the end you wind up up one return. If you want a more detailed example just let me know!